Fabric.Identity.API/Extensions/AuthenticationBuilderExtensions.cs

```diff
@@ -0,0 +1,124 @@
+                    options.SignOutScheme = IdentityServerConstants.SignoutScheme;
+                    options.Authority = appConfiguration.AzureActiveDirectorySettings.Authority;
+                    options.ResponseType = OpenIdConnectResponseType.CodeIdToken;
+                    options.ClaimsIssuer = appConfiguration.AzureActiveDirectorySettings.ClaimsIssuer;
+                    options.ClientId = appConfiguration.AzureActiveDirectorySettings.ClientId;
+                    options.ClientSecret = appConfiguration.AzureActiveDirectorySettings.ClientSecret;
+                    options.CallbackPath = "/signin-oidc-" + FabricIdentityConstants.AuthenticationSchemes.Azure;
+                    options.SignedOutCallbackPath = "/signout-callback-oidc-" + FabricIdentityConstants.AuthenticationSchemes.Azure;
+                    options.GetClaimsFromUserInfoEndpoint = true;
+                    // aspnetcore 2.2 removes most claims by default but we need the claims issuer.
+                    // In order to add them back you have to do the following
+                    options.ClaimActions.Remove("iss");
+                    options.TokenValidationParameters = new TokenValidationParameters
+                    {
+                        ValidateIssuer = false
+                    };
+
+                    foreach (var s in appConfiguration.AzureActiveDirectorySettings.Scope)
+                    {
+                        options.Scope.Add(s);
+                    }
+
+                });
+        }
+
+        public static AuthenticationBuilder AddExternalIdentityProviders(this AuthenticationBuilder builder,
+            IAppConfiguration appConfiguration)
+        {
+            if (appConfiguration.ExternalIdProviderSettings?.ExternalIdProviders?.Any() == true)
+            {
+                // Add OpenIdConnect options to each provider.
+                foreach (var externalIdProvider in appConfiguration.ExternalIdProviderSettings.ExternalIdProviders)
+                {
+                    builder.AddOpenIdConnect(externalIdProvider.ProviderName, externalIdProvider.DisplayName, options =>
+                    {
+                        options.SignInScheme = IdentityServerConstants.ExternalCookieAuthenticationScheme;
+                        options.SignOutScheme = IdentityServerConstants.SignoutScheme;
+
+                        options.Authority = externalIdProvider.Authority;
+                        options.ClientId = externalIdProvider.ClientId;
+                        options.ClientSecret = externalIdProvider.ClientSecret;
+                        options.ResponseType = externalIdProvider.ResponseType;
+                        options.CallbackPath = "/signin-oidc-" + externalIdProvider.ProviderName;
+                        options.SignedOutCallbackPath = "/signout-callback-oidc-" + externalIdProvider.ProviderName;
+                        options.SaveTokens = true; // Okta needs the id_token for successful logout
```

Comment:
Does this need to be a configuration value per external IdP? Okta needs it to be true, but could others need it false? I guess I'd be fine if we keep it true for now and let usage show us if we need to add it as a config.

Reply:
The external IdPs don't need it, in my testing, it could have an if block, but we don't really know what they will call it for the providername, since the customer has to name it.